### PR TITLE
feat: add typescript:update command

### PR DESCRIPTION
**Pull request by mdonnalley**
### What does this PR do?
- Adds `typescript:update` command which does the following:
    - updates typescript version to latest
    - updates ES target to ESNext
    - installs dependencies
    - runs tests

- renames `ProjectJson` interface to `PackageJson`, which is more accurate

### What issues does this PR fix or reference?
@W-8667963@

### CODE REVIEW

File: src/commands/typescript/update.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    if (/ES[0-9]{4}/g.test(this.flags.target)) return true;
+
+    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+    throw new SfdxError(`Invalid target: ${this.flags.target}`, 'InvalidTargetVersion');
+  }
+
+  private validateTsVersion(typescriptPkg: NpmPackage): boolean {
+    if (this.flags.version) {
+      if (!typescriptPkg.versions.includes(this.flags.version)) {
```
Comment (WillieRuemmele) [inline src/commands/typescript/update.ts:95]: Any reason to not combine these into one if statement? Same with line 85+87 above

File: src/commands/typescript/update.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    if (pkg.packageJson.devDependencies['typescript']) {
+      pkg.packageJson.devDependencies['typescript'] = newVersion;
+    }
+
+    // If the prepare script runs sf-install, the install will fail because the typescript version
+    // won't match the expected version. So in that case, we delete the prepare script so that we
+    // get a successful install
+    if (pkg.packageJson.scripts['prepare'] === 'sf-install') {
+      delete pkg.packageJson.scripts['prepare'];
+    }
```
Comment (WillieRuemmele) [inline src/commands/typescript/update.ts:70]: Do we ever need to put this back into the package?

**Review by WillieRuemmele — APPROVED** [review]

File: src/commands/typescript/update.ts (hunk 2 above)
Comment (mdonnalley) [reply src/commands/typescript/update.ts:70]: @WillieRuemmele No, these changes aren't being committed - we just need them so we can run the tests in circle with the updated ts version

**Review by peternhale — APPROVED** [review]

File: src/package.ts [hunk 3]
```diff
@@ -53,9 +57,9 @@ export class Package extends AsyncOptionalCreatable {
     this.location = location;
   }
 
-  public async readProjectJson(): Promise<ProjectJson> {
+  public async readPackageJson(): Promise<PackageJson> {
     const pkgJsonPath = this.location ? path.join(this.location, 'package.json') : 'package.json';
-    return (await fs.readJson(pkgJsonPath)) as ProjectJson;
+    return (await fs.readJson(pkgJsonPath)) as PackageJson;
```
Comment (shetzel) [inline src/package.ts:62]: This is essentially the same as the exported function above.

1. should we reuse 1 function?
2. should we just make it a synchronous read (readJsonSync) or is async more appropriate?

File: src/commands/typescript/update.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+  private async updateEsTarget(): Promise<void> {
+    const tsConfigPath = path.resolve('tsconfig.json');
+    const tsConfig = await fs.readJson(tsConfigPath);
+    set(asObject(tsConfig), 'compilerOptions.target', this.flags.target);
+    this.ux.log(`Updating tsconfig target at ${tsConfigPath} to:`, this.flags.target);
+    await fs.writeJson(tsConfigPath, tsConfig);
+  }
+
+  private async updateTsVesion(typescriptPkg: NpmPackage): Promise<void> {
```
Comment (shetzel) [inline src/commands/typescript/update.ts:56]: spelling --> `updateTsVersion`

File: src/commands/typescript/update.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+      description: messages.getMessage('typescriptVersion'),
+    }),
+    target: flags.string({
+      char: 't',
+      description: messages.getMessage('esTarget'),
+      default: 'ESNext',
+    }),
+  };
+
+  public async run(): Promise<void> {
```
Comment (shetzel) [inline src/commands/typescript/update.ts:34]: Would it be helpful to return a result?  Will this be used from CI and the result could be helpful in the reporting?

File: src/commands/typescript/update.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      throw new SfdxError('Could not find typescript on the npm registry', 'TypescriptNotFound');
+    }
+  }
+
+  private validateEsTarget(): boolean {
+    if (this.flags.target === 'ESNext') return true;
+
+    if (/ES[0-9]{4}/g.test(this.flags.target)) return true;
+
+    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
```
Comment (shetzel) [inline src/commands/typescript/update.ts:89]: I think we can remove this disabling if we use `SfdxError.create()` and put the error messages in the message file.

File: src/package.ts [hunk 7]
```diff
@@ -100,12 +104,12 @@ export class Package extends AsyncOptionalCreatable {
...
   public writePackageJson(rootDir?: string): void {
     const pkgJsonPath = rootDir ? path.join(rootDir, 'package.json') : 'package.json';
-    fs.writeJsonSync(pkgJsonPath, this.projectJson);
+    fs.writeJsonSync(pkgJsonPath, this.packageJson);
   }
 
   public pinDependencyVersions(targetTag: string): ChangedPackageVersions {
     // get the list of dependencies to hardcode
-    const dependencies: string[] = this.projectJson['pinnedDependencies'];
+    const dependencies: string[] = this.packageJson['pinnedDependencies'];
```
Comment (shetzel) [inline src/package.ts:112]: We should have doc for all public APIs.

**Review by shetzel — CHANGES_REQUESTED** [review]

File: src/commands/typescript/update.ts (hunk 2 above)
Comment (amphro) [reply src/commands/typescript/update.ts:70]: What about running these locally? I think we should make it clear (output to the use at the end of the run) that any file changes from this command should NOT be committed.

Comment (amphro) [reply src/commands/typescript/update.ts:70]: i.e. This is for testing new versions only and to actually update the version it must go through dev-scripts.

File: src/package.ts (hunk 7 above)
Comment (amphro) [reply src/package.ts:112]: Agreed

File: src/repository.ts [hunk 8]
```diff
@@ -113,6 +113,10 @@ abstract class Repository extends AsyncOptionalCreatable {
     this.execCommand('yarn build', silent);
   }
 
+  public test(silent = false): void {
```
Comment (amphro) [inline src/repository.ts:116]: When would someone want this silent? WOuldn't you always want to see the test output? Is it for unit test?

**Pull request merged**